site/docs/spec/terminology.yaml

```diff
@@ -0,0 +1,230 @@
+  defined along with the parameter.
+enum function parameter: >- # changed from "enumeration"
+  a {function parameter} to which an {enum function argument} can be bound.
+  The allowed values are defined along with the parameter by means of a set
+  of string options and an optional default option.
+
+function binding: >- # didn't really exist
+  information about how a {function} is to be called, defined in the {plan}.
+  Includes at least a function {reference} and the {function argument}s bound
+  to the function.
+function argument: >- # didn't really exist
+  the information bound to a {function parameter} as part of a
+  {function binding}. Can be a {value function argument},
+  {type function argument}, or {enum function argument}.
+value function argument: >- # didn't really exist
+  a {function argument} consisting of a {static} {datum} or a {runtime}
+  {expression} (evaluated for each invocation of the function) passed to a
+  {value function parameter}.
+type function argument: >- # didn't really exist
+  a {static} {data type} passed to a {type function parameter}.
+enum function argument: >- # didn't really exist
+  a {static} {enum option} passed to an {enum function parameter}.
+
+enum option: >- # also just "enumeration", ambiguous with the parameter type
+  a special type of value for {function argument}s, used to pass {static}
+  configuration information to {function}s in a type-safe way.
+
+data type: >- # was just "type", using "data type" to distinguish from metatypes
+  a concrete type as defined [here](../types/type_system.md), consisting of
+  a {type class}, a {nullability flag}, a {type variation}, and zero or more
+  {type argument}s as defined by the {type class}.
+
+type class: >-
+  a non-concrete type as defined [here](../types/type_classes.md).
+  Subdivided into {simple type class}es, {compound type class}es, and
+  {user-defined} type classes.
+simple type class: >-
+  a {type class} defined with an empty {type parameter pack}.
+compound type class: >-
+  a {type class} defined with an non-empty {type parameter pack}.
+
+nullability flag: >-
+  specifies whether a {data type} includes the special {null} value. Either
+  {nullable} or {non-nullable}.
+nullable: >-
+  qualifier for a {data type} that signals that {datum}s of this type may be
+  {null}.
+non-nullable: >- # a.k.a. "required", but that confused me to no end initially
+  qualifier for a {data type} that signals that {datum}s of this type may
+  NOT be {null}.
+
+type variation: >-
+  a specific variation of a {type class}, allowing different variations of the
+  same {type class} to exist in a system at a time. Usually this is set to the
+  {system-preferred variation} for the {type class}.
+system-preferred variation: >-
+  a special {type variation} that is implicitly defined for each {type class},
+  serving as a default value for when {type variation}s are not in use.
+
+type parameter pack: >- # wasn't really used
```

Comment:
Couple of initial thoughts here on this and much below this:
- Trying to finalize things we agree on and don't know if agree on in the same patch might slow down the merging of this patch.
- You started using parameter when talking about functions. I basically said I'm only going to use parameter for type related stuff and arguments for function related stuff. This is a slight manipulation of the terms but avoids the need here and above of constantly having "function type parameter". By using both, it actually makes things much more confusing to me. Especially when you didn't use "type type" below. So now we have type argument here but we also have type arguments in functions.
- I think that you've collapsed data types and parameterized types together which I think is suboptimal. There are three independent type abstractions, one building on the next. Type literals, parameterized types and derivation expressions. They all deal with types, they are at just different levels of capability (expressibility). Each can only be used at a certain location within the simple extensions.
- One of the things that is especially confusing when reading your decomposition of this is what appears to be a mixing of compound types and parameterized types. Compound types are type literals that express no additional abstract concepts. For example, `VARCHAR<40>`. This is discrete from a parameterized type such as `VARCHAR<T>` or a type class e.g. `VARCHAR`. In a function declaration, I cannot use compound types for bindings (e.g. `VARCHAR<40>`). I can only parameterized types (e.g. `VARCHAR<T>`). The parameter can be used or ignored. In parameterized types I can also use the `anyX` pseudo-type parameter (X is an optional numeric identifier after any) which allows is a wildcard which is reusable. This is useful when you want to say something like `add(any1, any1) => any1`. This is a function which says I can "add" two values of any type to get a new version of that type. The any is a wildcard but it is the same wildcard everytime it is referenced.

Reply:
I will take a more thorough pass to your concepts in a bit.

Reply:
"Compound" as a descriptor for `VARCHAR<40>` seems like non-standard type system terminology.

`VARCHAR<40>` is as fully specified a type as `int32`. Calling `VARCHAR<40>` "compound" would be like saying `int32` is a compound version of `int<N>` (assume that exists for argument's sake).

> Compound types are type literals that express no additional abstract concepts.

By this definition every monomorphized type including any type that doesn't have a parameter (`int32`, `float64`, etc) is a compound type. If everything is a compound type it doesn't really make sense to keep it around as a distinct concept or term.

Reply:
FWIW, Substrait has used the term "simple" to refer to all types that don't need a `<...>` pack to be concrete, and "compound" to refer to all types that do, [since its conception](https://github.com/substrait-io/substrait/blob/c7f7ef80e70f5b758dfab604ee06dbfb4ddfe020/doc/types_logical_compound.md). But I agree. "Compound" feels synonymous to "nested" to me.